src/sdk/namespace_fs.js

```diff
@@ -199,6 +215,55 @@ const dir_cache = new LRUCache({
     max_usage: config.NSFS_DIR_CACHE_MAX_TOTAL_SIZE,
 });
 
+/**
+ * @typedef {{
+ *  time: number,
+ *  stat: fs.Stats,
+ *  usage: number,
+ *  sorted_entries?: fs.Dirent[],
+ * }} ReaddirVersionsCacheItem
+ * @type {LRUCache<object, string, ReaddirCacheItem>}
+ */
+const versions_dir_cache = new LRUCache({
+    name: 'nsfs-versions-dir-cache',
+    make_key: ({ dir_path }) => dir_path,
+    load: async ({ dir_path, fs_context }) => {
+        const time = Date.now();
+        const stat = await nb_native().fs.stat(fs_context, dir_path);
+        const sorted_entries = [];
+        let usage = config.NSFS_DIR_CACHE_MIN_DIR_SIZE;
+        if (stat.size <= config.NSFS_DIR_CACHE_MAX_DIR_SIZE) {
+            const sorted_latest_versions = await nb_native().fs.readdir(fs_context, dir_path);
+            /*eslint no-unused-expressions: ["error", { "allowTernary": true }]*/
+            sorted_latest_versions.sort(sort_entries_by_name);
+            /* list all versions of each entry inside the dir_path under .versions path */
+            const version_path = dir_path + "/" + HIDDEN_VERSIONS_PATH;
+            const sorted_versions = await nb_native().fs.readdir(fs_context, version_path);
+            sorted_versions.sort(sort_entries_by_time);
+            sorted_latest_versions.forEach(entry => {
+                sorted_entries.push(entry);
+                sorted_versions.forEach(version => {
+                    if (version.name.startsWith(entry.name)) {
+                        version.name = HIDDEN_VERSIONS_PATH + "/" + version.name;
+                        sorted_entries.push(version);
+                    }
+                });
+            });
```

Comment:
I think we can avoid this nested loop of entries and versions. The format of the version name allows us to sort them together

something like that (I may have syntax errors here):

```suggestion
            const latest_versions = await nb_native().fs.readdir(fs_context, dir_path);
            const version_path = dir_path + "/" + HIDDEN_VERSIONS_PATH;
            const old_versions = await nb_native().fs.readdir(fs_context, version_path);
            const entries = latest_versions.concat(old_versions);
            const sorted_entries = entries.sort(sort_entries_by_name_and_time);
```

Reply:
With the suggested change, the `version_path` in the sorted arrays will be wrongly set. i.e assume that
the latest file is in `$BUCKET/dir1/file1` and  its version path will be `$BUCKET/dir1/.versions/file1_mtime-*`

with the suggested change, we will have `readdir(dir1)` which gives proper path i.e `$BUCKET/dir1/file1` but
`readdir(`dir1/.versions`)` will lead to `$BUCKET/dir1/file1_mtime-*` (NOTE missing `.versions` in the path)

This is handled from line no: 247-248 in the above code

>                     if (version.name.startsWith(entry.name)) {
>                         version.name = HIDDEN_VERSIONS_PATH + "/" + version.name;
>                         sorted_entries.push(version);
>                     }
>

Reply:
Discussed and resolved